common/datasets/sms_wsj/returnn_datasets.py

```diff
@@ -0,0 +1,557 @@
+                [
+                    self._pad_label * np.ones(pad_start),
+                    hdf_classes[speaker_idx],
+                    self._pad_label * np.ones(pad_end),
+                ]
+            )
+        d["target_classes"] = np.stack(hdf_classes).T
+        d["target_classes_len"] = np.array(padded_len)
+        return d
+
+    def get_seq_length_for_keys(self, seq_idx: int) -> NumbersDict:
+        """
+        Returns sequence length for all data/target keys.
+        """
+        d = super().get_seq_length_for_keys(seq_idx)
+        data = self[seq_idx]
+        d["target_classes"] = int(data["target_classes_len"])
+        return NumbersDict(d)
+
+
+class SmsWsjWrapper(MapDatasetWrapper):
+    """
+    Base class for datasets that can be used in RETURNN config.
+    """
+
+    def __init__(self, sms_wsj_base, **kwargs):
+        """
+        :param SmsWsjBase sms_wsj_base: SMS-WSJ base class to allow inherited classes to modify this
+        """
+        if "seq_ordering" not in kwargs:
+            print("Warning: no shuffling is enabled by default", file=returnn_log.v2)
+        super().__init__(sms_wsj_base, **kwargs)
+        # self.num_outputs = ...  # needs to be set in derived classes
+
+        def _get_seq_length(seq_idx: int) -> NumbersDict:
+            """
+            Returns sequence length for all data/target keys.
+            """
+            corpus_seq_idx = self.get_corpus_seq_idx(seq_idx)
+            return sms_wsj_base.get_seq_length_for_keys(corpus_seq_idx)
+
+        self.get_seq_length = _get_seq_length
+
+    @staticmethod
+    def _pre_batch_transform(inputs: Dict[str, Any]) -> Dict[str, np.array]:
+        """
+        Used to process raw SMS-WSJ data
+        :param Dict[str, Any] inputs: input as coming from SMS-WSJ
+        """
+        return_dict = {
+            "seq_tag": np.array(inputs["example_id"], dtype=object),
+            "source_id": np.array(inputs["source_id"], dtype=object),
+            "seq_len": np.array(inputs["num_samples"]["observation"]),
+        }
+        return return_dict
+
+    def _collect_single_seq(self, seq_idx: int) -> DatasetSeq:
+        """
+        :param seq_idx: sorted seq idx
+        """
+        corpus_seq_idx = self.get_corpus_seq_idx(seq_idx)
+        self._dataset.update_buffer(corpus_seq_idx)
+        data = self._dataset[corpus_seq_idx]
+        assert "seq_tag" in data
+        return DatasetSeq(seq_idx, features=data, seq_tag=data["seq_tag"])
+
+    def init_seq_order(self, epoch: Optional[int] = None, **kwargs) -> bool:
+        """
+        Override this in order to update the buffer. get_seq_length is often called before _collect_single_seq,
+        therefore the buffer does not contain the initial indices when continuing the training from an epoch > 0.
+        """
+        out = super().init_seq_order(epoch=epoch, **kwargs)
+        buffer_index = ((epoch or 1) - 1) * self.num_seqs % len(self._dataset)
```

Comment:
I think this logic may be buggy. I'm having trouble re-starting a training that crashed at a later sub-epoch because the `seq_idx` requested in `get_seq_length` is not present in the buffer. I think this calculation here assumes that `self.num_seqs` is the same for all sub-epochs which may be incorrect? If the total sequence count is not divisible by `partition_epoch` then `self.num_seqs` may differ between subepochs which can accumulate here when `epoch` is high.

Reply:
Maybe you can try
```suggestion
        buffer_index = ((epoch or 1) - 1) % (self.partition_epoch or 0) * self.num_seqs
```
Here, the error does not accumulate much but at maximum `self.partition_epoch` times.